### PR TITLE
Make LArray supporting pandas >= 0.21 (issue 569) 

### PR DESCRIPTION
**Pull request by alixdamman**
Needs a changelog entry. Waiting to know if we merge this PR before or after the release of the 0.29 version

### CODE REVIEW

File: larray/inout/pandas.py [hunk 1]
```diff
@@ -239,8 +248,11 @@ def df_aslarray(df, sort_rows=False, sort_columns=False, raw=False, parse_header
                 # (see https://github.com/pandas-dev/pandas/issues/15105)
                 df = series.unstack(level=-1, fill_value=fill_value)
                 # pandas (un)stack and pivot(_table) methods return a Dataframe/Series with sorted index and columns
-                labels = df_labels(series, sort=False)
-                index = pd.MultiIndex.from_tuples(list(product(*labels[:-1])), names=series.index.names[:-1])
+                labels = index_to_labels(series.index, sort=False)
+                if isinstance(df.index, pd.core.index.MultiIndex):
```
Comment (gdementen) [inline larray/inout/pandas.py]: This code is getting really ugly. Couldn't some of this code be transferred to from_series to support multiIndex and here only use something like:
```python
series = df[df.columns[-1]]
return from_series(sort_rows=sort_columns)
```

File: larray/inout/pandas.py [hunk 2]
```diff
@@ -39,34 +39,39 @@ def parse(s):
...
+    # since version 0.21 of pandas, the MultiIndex constructors no longer squeezes
+    # a MultiIndex with all length-one levels down to a regular Index.
+    if isinstance(idx, pd.core.index.MultiIndex):
+        if sort_rows:
+            new_index = pd.MultiIndex.from_product(labels)
+        else:
+            new_index = pd.MultiIndex.from_tuples(list(product(*labels)))
     else:
-        new_index = pd.MultiIndex.from_tuples(list(product(*labels)))
+        new_index = idx
```
Comment (gdementen) [inline larray/inout/pandas.py]: I guess `sort_rows=True` is not respected in this case.

File: larray/core/array.py [hunk 3]
```diff
@@ -1002,8 +1002,12 @@ def to_frame(self, fold_last_axis_name=False, dropna=None):
                 tmp = axes_names[-1] if axes_names[-1] is not None else ''
                 if self.axes[-1].name:
                     axes_names[-1] = "{}\\{}".format(tmp, self.axes[-1].name)
-
-            index = pd.MultiIndex.from_product(self.axes.labels[:-1], names=axes_names)
+            # since version 0.21 of pandas, the MultiIndex constructors no longer squeezes
```
Comment (gdementen) [inline larray/core/array.py]: Unsure we need the comment. If we change the code by inadvertence, we will see tests failing, right?

File: larray/inout/pandas.py [hunk 4]
```diff
@@ -39,34 +39,39 @@ def parse(s):
...
+        return [list(idx.values)]
 
 
 def cartesian_product_df(df, sort_rows=False, sort_columns=False, **kwargs):
-    labels = df_labels(df, sort=sort_rows)
-    if sort_rows:
-        new_index = pd.MultiIndex.from_product(labels)
+    idx = df.index
+    labels = index_to_labels(idx, sort=sort_rows)
+    # since version 0.21 of pandas, the MultiIndex constructors no longer squeezes
```
Comment (gdementen) [inline larray/inout/pandas.py]: unsure we need the comment

File: larray/inout/pandas.py [hunk 5]
```diff
@@ -39,34 +39,39 @@ def parse(s):
...
+    # a MultiIndex with all length-one levels down to a regular Index.
```
Comment (alixdamman) [inline larray/inout/pandas.py]: This comment is no longer true...

File: larray/inout/pandas.py (hunk 2 above)
Comment (alixdamman) [reply larray/inout/pandas.py]: All tests are all green. This means, I guess, that a new test with `sort_rows` must be added to cover this case.

File: larray/inout/pandas.py [hunk 6]
```diff
@@ -74,27 +74,70 @@ def cartesian_product_df(df, sort_rows=False, sort_columns=False, **kwargs):
...
+        # pandas (un)stack and pivot(_table) methods return a Dataframe/Series with sorted index and columns
+        if not sort_rows:
+            labels = index_to_labels(s.index, sort=False)
+            if isinstance(df.index, pd.core.index.MultiIndex):
+                index = pd.MultiIndex.from_tuples(list(product(*labels[:-1])), names=s.index.names[:-1])
+            else:
+                index = labels[0]
+            columns = labels[-1]
+            df = df.reindex(index=index, columns=columns, fill_value=fill_value)
+        return from_frame(df, sort_rows=sort_rows, sort_columns=sort_rows, meta=meta, **kwargs)
```
[the remark this answers is not in the document]
Comment (alixdamman) [reply larray/inout/pandas.py]: from_frame() does not use from_series()

[the remark this answers is not in the document]
Comment (gdementen) [reply larray/inout/pandas.py]: sorry, I misread the diff. df_aslarray uses from_series, not from_frame

File: doc/source/changes/version_0_29.rst.inc [hunk 7]
```diff
@@ -375,3 +375,5 @@ Fixes
 
 * fixed taking a subset of an array by giving an index along a specific axis using a string (strings
   like 'axisname.i[pos]').
+
+* made LArray supporting pandas versions >= 0.21 (closes :issue:`569`)
```
Comment (gdementen) [inline doc/source/changes/version_0_29.rst.inc]: s/supporting/support/

File: larray/inout/pandas.py [hunk 8]
```diff
@@ -39,61 +39,113 @@ def parse(s):
...
     if prodlen == len(df) and columns == list(df.columns) and np.array_equal(df.index.values, new_index.values):
         return df, labels
-    return df.reindex(new_index, columns, **kwargs), labels
+    return df.reindex(index=new_index, columns=columns, fill_value=fill_value, **kwargs), labels
 
 
-def from_series(s, sort_rows=False):
-    """
-    Converts Pandas Series into 1D LArray.
+def from_series(s, sort_rows=False, fill_value=nan, meta=None, **kwargs):
```
Comment (gdementen) [inline larray/inout/pandas.py]: the support for multiIndex in from_series needs to be in the release notes I think. fill_value and meta can be omitted from them IMO.

**Pull request merged**